Generics.md

```diff
@@ -0,0 +1,164 @@
+  }
+  ```
+
+Here’s a more complicated case. reduce is parameterized by an element type (e.g. u32) and a binary operation, e.g. max()
+
+* ```
+    ./util.wgsl:
+
+    export<E, BinOp> 
+    fn reduce(elems: array<E, 2>) -> E { 
+      return BinOp(elems[0], elems[1]); 
+    }
+
+* ```
+    ./main.wgsl:
+
+    import ./ops/binOpMax<f32> as binOpMax;
+    import ./util/reduce<f32, binOpMax> as maxF32; 
+
+    fn main() {
+      maxF32(a1);
+    }
+    ```
+
+Note that you can import a generic function w/o providing parameters:
+
+* ```
+    ./util.wgsl:
+
+    import binOpMax from ./ops;   // no generic variable specified yet
+
+    export fn reduceMax<E>(elems: array<E, 2>) -> E { 
+      return binOpMax<E>(elems[0], elems[1]); // generic value applied at call site
+    }
+    ```
+
+Re-exporting generics is allowed (presuming we allow re-exporting in general, see [Visibility](./Visiblity.md)):
+
+* ```
+    ./lib.wgsl:
+    export reduce from util/reduce.wgsl; // re-export at package root level
+
+    ./util/reduce.wgsl:
+    export<E, BinOp> fn reduce(elems: array<E, 2>) -> E { }
+    ```
+
+## Questions and possible extensions
+
+* Do angle brackets conflict or comport with wgsl templates?
+* Can you export a generic function after variable substitution too? or only the generic version
+* Allow generic values to be pulled from runtime parameters?
+  wgsl-linker currently recognizes an `ext.` prefix to get variable values from the runtime caller.
+  e.g. ext.workgroupSize would patch in runtime variables at link time.
+  Hopefully we can address that with runtime #define, we’ll see.
+* Currently there are no type constraints available for generic variable declarations..
+  Simply substituting parameters and letting dawn or naga typecheck at runtime seems ok for now.
+  A future type checker could check annotation uses are valid by substituting generic parameters
+  and type checking the expanded wgsl.
+  And of course a future version of wgsl or wesl generics could add explicit type constraints on generic variables.
+
+* Generics on structs too?
+
+  * ```
+      ./util.wgsl:
+      export struct Point<T> { position: vec2<T>, color vec3f } 
+
+      ./main.wgsl
+      import Point<u32> as UPoint from ./util
+
+      fn main() {
+        let p = UPoint(vec2u(0, 0), vec3f(.5, .5, .5));
+      }
+      ```
+
+* The reduce example makes me think whether we could call a generic function recursively,
+  and what would that do.
+  In theory, the following would unroll to N nested function calls.
+  The wgsl compilers may be good at flattening this.
+
+* ```
+    fn accumulate<E, Op, N>(acc: E, elems: array<E, N>) -> E {
+      if N >= 2 {
+        return accumulate<E, Op, N-1>(accumulateBinOp(E, elems[N-1]), elems);
+      else {
+        return acc;
+      }
+    }
+
+    fn op_add<E>(e1: E, e2: E) {
+      return e1 + e2;
+    }
+
+    fn array_sum<E, N>(elems: array<E, N>) -> E {
+      return accumulate<E, op_add<E>, N>(0, elems);
```

Comment:
Could you not get around that by doing a cast using to `E()`?

Reply:
aah yes, that makes sense :)

Reply:
yup, good find :)